src/backend/cdb/dispatcher/cdbdisp_query.c

```diff
@@ -1676,3 +1681,104 @@ findParamType(List *params, int paramid)
 
 	return InvalidOid;
 }
+
+/*
+ * process data in pointer cosume_p, ex: copy everything interested in.
+ * For EP_TAG_I, EP_TAG_U, EP_TAG_D, consume_p is a Bitmapset**, we just
+ * copy the content and process them later.
+ */
+void
+ConsumeExtendProtocolData(ExtendProtocolSubTag subtag, void *consume_p)
+{
+	Assert(epd);
+
+	if ((epd->consumed_bitmap & (1 << subtag)) == 0)
+		return;
+
+	switch (subtag)
+	{
+		case EP_TAG_I:
+		case EP_TAG_U:
+		case EP_TAG_D:
+			Assert(consume_p != NULL);
+			*((Bitmapset **) consume_p) = bms_copy(list_nth(epd->subtagdata, subtag));
+			bms_free(list_nth(epd->subtagdata, subtag)); /* clean up */
+			break;
+		default:
+			Assert(false);
+	}
+
+	/* Mark subtag consumed. */
+	epd->consumed_bitmap &= ~(1 << subtag);
+}
+
+/*
+ * Contents must be allocated in TopTransactionMemoryContext.
+ */
+void InitExtendProtocolData(void)
+{
+	MemoryContext oldctx = MemoryContextSwitchTo(TopTransactionContext);
```

Comment:
ExtendProtocol data struct could be more general ？

Reply:
The `ExtendProtocol` data struct is designed to allow users to define their own extended subtag data list, so it’s flexible based on their specific needs. 

That said, agree that the struct could be more general, and we do have plans to refactor it, along with `libpq`, to improve its versatility. However, these changes won’t be included in this PR and will be addressed in a future update.

Reply:
I suggest use json or jsonb to more general.

Reply:
Eh...embed json/jsonb inside kernel?  Surprised to hear that.

Reply:
> Eh...embed json/jsonb inside kernel? Surprised to hear that.

https://www.postgresql.org/docs/9.5/functions-json.html

Reply:
> > Eh...embed json/jsonb inside kernel? Surprised to hear that.
> 
> https://www.postgresql.org/docs/9.5/functions-json.html

So what and why?

Reply:
to make extend protocol of libpq schema less.

Reply:
No such need, it makes no sense to make things complicated. 
And json/jsonb is a database object, it's very stupid to make them embedding inside internal  codes.